enterprise/server/invocation_stat_service/invocation_stat_service.go

```diff
@@ -485,6 +777,211 @@ func (i *InvocationStatService) GetInvocationStat(ctx context.Context, req *inpb
+	// invocations / executions in each group to come first in the results so that
+	// we can use it as an input into how we rank the charts.  To do this, we
+	// toss in a bogus column and sort by it to ensure that totals come first.
+	if col == "" {
+		return fmt.Sprintf(
+			`(SELECT %s, 0 AS totals_first, count(*) AS total,
+					countIf(%s) AS selection, countIf(not(%s)) AS inverse
+				FROM %s %s)`,
+			nulledOutFieldList, drilldown, drilldown, table, where), args
+	}
+
+	return fmt.Sprintf(`
+		(SELECT %s, 1 AS totals_first, count(*) AS total, countIf(%s) AS selection,
+			countIf(not(%s)) AS inverse
+		FROM %s %s
+		GROUP BY %s ORDER BY selection DESCENDING, total DESCENDING LIMIT 25)`,
+		nulledOutFieldList, drilldown, drilldown, table, where, col), args
+}
+
+func getDrilldownQueryFilter(filters []*sfpb.StatFilter) (string, []interface{}, error) {
+	if len(filters) == 0 {
+		return "", nil, status.InvalidArgumentError("Empty filter for drilldown.")
+	}
+	var result []string
+	var resultArgs []interface{}
+	for _, f := range filters[:] {
+		str, args, err := filter.GenerateFilterStringAndArgs(f, "")
+		if err != nil {
+			return "", nil, err
+		}
+		result = append(result, str)
+		resultArgs = append(resultArgs, args...)
+	}
+	return strings.Join(result, " AND "), resultArgs, nil
+}
+
+// TODO(jdhollen): This can be made much efficient using GROUPING SETS when we
+// are able to upgrade to clickhouse 22.6 or later.  The release date for 22.8
+// from Altinity is supposed to be 2023-02-15.
+func (i *InvocationStatService) getDrilldownQuery(ctx context.Context, req *stpb.GetStatDrilldownRequest) (string, []interface{}, error) {
+	drilldownFields := []string{"user", "host", "pattern", "repo_url", "branch_name", "commit_sha"}
+	placeholderQuery := query_builder.NewQuery("")
+
+	if err := addWhereClauses(placeholderQuery, req.GetQuery(), req.GetRequestContext(), 0); err != nil {
+		return "", nil, err
+	}
+
+	drilldownStr, drilldownArgs, err := getDrilldownQueryFilter(req.GetFilter())
+	if err != nil {
+		return "", nil, err
+	}
+
+	whereString, whereArgs := placeholderQuery.Build()
+
+	args := make([]interface{}, 0)
+	queries := make([]string, 0)
+	subQStr, subQArgs := i.getDrilldownSubquery(ctx, drilldownFields, req, whereString, whereArgs, drilldownStr, drilldownArgs, "")
+	queries = append(queries, subQStr)
+	args = append(args, subQArgs...)
+	for _, s := range drilldownFields {
+		subQStr, subQArgs := i.getDrilldownSubquery(ctx, drilldownFields, req, whereString, whereArgs, drilldownStr, drilldownArgs, s)
+		queries = append(queries, subQStr)
+		args = append(args, subQArgs...)
+	}
+
+	return fmt.Sprintf("SELECT * FROM (%s) ORDER BY totals_first", strings.Join(queries, " UNION ALL ")), args, nil
+}
+
+func addOutputChartEntry(m *map[inpb.AggType]*stpb.DrilldownChart, dm *map[inpb.AggType]float64, aggType inpb.AggType, label *string, inverse int64, selection int64, totalInBase int64, totalInSelection int64) {
```

Comment:
maps types are references so you don't need to pass them with a pointer in go (which should simplify a lot of the code below)

Reply:
Thanks, I went wild with these before I realized what the deal was (strings too, you should have seen it), and got most things but missed this.